src/test/cpp/subsystems/swerve_drive_test.cpp

```diff
@@ -0,0 +1,168 @@
+    swerve_->Periodic();
+
+    // drive motors shouldn't be 0 then (not sure exactly what they should be, but it shouldn't be 0)
+    EXPECT_NE(sw_interface_.left_front_drive_motor_output, 0);
+    EXPECT_NE(sw_interface_.left_back_drive_motor_output, 0);
+    EXPECT_NE(sw_interface_.right_front_drive_motor_output, 0);
+    EXPECT_NE(sw_interface_.right_back_drive_motor_output, 0);
+    //FIXME
+
+    //TODO steer motors
+}
+
+TEST_F(SwerveDriveTest, LockAutonomousDrive) {
+    //TODO better pos/heading
+    frc::Pose2d pos = frc::Pose2d();
+
+    ASSERT_TRUE(swerve_->InitAuto(pos, true)); // lock heading
+
+    ASSERT_TRUE(swerve_->RunAuto());
+
+    //do some janky periodic() stuff maybe? just get it to go through all the states
+    //TODO
+}
+
+TEST_F(SwerveDriveTest, UnlockedAutoDrive) {
+     //TODO better pos/heading
+    frc::Pose2d pos = frc::Pose2d();
+
+    ASSERT_TRUE(swerve_->InitAuto(pos, false)); // turn to heading first before driving
+
+    ASSERT_TRUE(swerve_->RunAuto());
+
+    //do some janky periodic() stuff maybe? just get it to go through all the states
+    //TODO
+}
+
+TEST_F(SwerveDriveTest, GyroTest) {
+    // initialize yaw in interface
+    sw_interface_.imu_yaw = 0;
+
+    // check to make sure the gyro can be reset
+    ASSERT_TRUE(swerve_->ResetGyro());
+    EXPECT_EQ(sw_interface_.reset_gyro, true);
+
+    // check to make sure the software interface updates
+    // Eli's number - thank you Eli
+    const double yaw = 64;
+    sw_interface_.imu_yaw = yaw;
+    EXPECT_EQ(sw_interface_.imu_yaw, yaw);
+
+    // check to make sure software interface updates make it to the subsystem
+    double swerve_yaw = 0.0;
+    ASSERT_TRUE(swerve_->GetHeading(&swerve_yaw));
+    EXPECT_EQ(yaw, swerve_yaw);
+}
+
+TEST_F(SwerveDriveTest, DriveEncodersTest) {
+    const double left_front = -100;
+    const double left_back = -50;
+    const double right_front = 100;
+    const double right_back = 200;
+
+    ASSERT_TRUE(swerve_->ResetDriveEncoders());
+    EXPECT_EQ(sw_interface_.reset_drive_encoders, true);
+    
+    sw_interface_.left_front_drive_motor_enc = left_front;
+    sw_interface_.left_back_drive_motor_enc = left_back;
+    sw_interface_.right_front_drive_motor_enc = right_front;
+    sw_interface_.right_back_drive_motor_enc = right_back;
+    
+    double avg = 0.0;
+    ASSERT_TRUE(swerve_->GetLeftDriveEncoderAverage(&avg));
+    EXPECT_EQ(avg, (left_front+left_back) / 2);
+    
+    ASSERT_TRUE(swerve_->GetRightDriveEncoderAverage(&avg));
+    EXPECT_EQ(avg, (right_front+right_back) / 2);
+
+    ASSERT_TRUE(swerve_->GetDriveEncoderAverage(&avg));
+    EXPECT_EQ(avg, (left_front+left_back+right_front+right_back) / 4);
+
+    //TODO velocity encoders
+    //FIXME this test fails
+}
+
+TEST_F(SwerveDriveTest, SteerEncodersTest) {
+    //TODO not sure what to test. this might be more of a SwerveModule than a SwerveDrive thing
+}
+
+TEST_F(SwerveDriveTest, PeriodicTest) {
+    // wait how does this test pass?
```

Comment:
This passes because you aren't checking anything with ASSERT or EXPECT

Reply:
but I figured if it threw an exception somewhere it should fail the test right?

Reply:
we dont throw exceptions though - we only return true early if a check or call fails. so the function just exits gracefully.

if you had `ASSERT_TRUE(Periodic());` you coukd catch the failure